File: backend/app/services/video_checkpoints.py
"""Checkpoints & LoRAs of a VIDEO dataset — both lanes in one list, and every
verb the image workspace's Checkpoints section has, at the unit of the STEP.

WHY THE STEP AND NOT THE FILE. A Wan 2.2 checkpoint is two files at one step
(`_high_noise` / `_low_noise`), and every operation that takes a file where it
should take a step gets it wrong without raising: a deploy that puts one expert
in ComfyUI, a delete that leaves the other half on disk, a "continue" that seeds
half a LoRA. So the list is grouped by step (`cloud_video_training.harvested_steps`
already does this for a cloud run; `group_saves_by_step` is that grouping made
shareable with the local folder), and deploy / delete take a step and act on
all of its files. Download stays per file, because two files side by side is
what ai-toolkit wrote and what every loader downstream expects.

WHY ONE LIST FOR TWO LANES. The image workspace shows local saves and cloud
runs' saves in the same section, with the same verbs. The video lane had the
cloud half only, rendered at the tail of the training block with a download link
per file and nothing else — a user who learned "📦 Deploy" on an image dataset
found no such button here (CLAUDE.md's parity rule; maintainer, 2026-09-02).

Deploy goes through the Video Test Studio's own copy (`video_test_studio`):
one folder, one naming, so a LoRA deployed from here is exactly what the
Studio's picker lists as "deployed". Deletes go through the app's trash like the
clips' removal does (`video_bank_service.remove_dataset_clips`) — never
`os.remove` — and a cloud run's per-file delete reuses
`cloud_training.delete_cloud_checkpoint`, which also clears the run's
`checkpoint_local_path` when it pointed at the trashed file.
"""
import json
import os

from ..extensions import db
from ..models import CloudTrainingRun, VideoDataset
from . import cloud_run_dataset as crd
from . import cloud_training as ct
from . import trash
from . import video_test_studio as vts
from . import video_training_local as vtl
from . import video_training

# Same destination as a removed clip: the app's own trash (Settings ▸ Storage),
# recoverable until it is emptied. Named on every delete answer so the UI's
# wording comes from `utils/deletionWording.js`, never from a hardcoded sentence.
DELETE_MODE = 'app_trash'

# What "ⓘ Details" shows of a cloud run's stamped parameters. An allow-list
# rather than the whole `train_params` blob: `resume_ckpt_paths` are pod-side
# machine paths and `auth_token`-class fields must never leave the row.
_PUBLIC_PARAMS = ('train_type', 'steps', 'base_model', 'low_vram', 'do_i2v',
                  'distillation', 'target_profile', 'frames', 'requested_gpu',
                  'resume_step', 'parent_run_id', 'auto_retry_of',
                  'sample_prompts')



def _group_saves_by_step(saves, target=None) -> list:
    """Cut a folder of saves into STEPS, so a Wan 2.2 pair travels together.

    DIVERGENCE 4 — upstream keeps this in `cloud_video_training.py` and its own
    docstring says the local run folder uses it too, "so both lanes cut a Wan
    pair at the same seam". That module is the rented-pod lane and is not
    carried here; the seam is local, so it lives with the lane that still
    exists. `target` numbers the FINAL save; the local lane passes None and the
    final is reported with `step: None`.
    """
    by_step = {}
    for name, path in saves.items():
        step, _stage = video_training.split_checkpoint_name(name)
        final = step is None
        key = (target if final else step, final)
        by_step.setdefault(key, []).append((name, path))
    out = []
    for (step, final), items in sorted(by_step.items(),
                                       key=lambda kv: (kv[0][1], kv[0][0] or 0)):
        items.sort()
        out.append({'step': None if step is None else int(step), 'final': final,
                    'files': [n for n, _ in items],
                    'paths': [p for _, p in items]})
    return out

def _dataset(user_id, dataset_id):
    ds = VideoDataset.query.filter_by(id=int(dataset_id), user_id=user_id).first()
    if ds is None:
        raise LookupError('video dataset not found')
    return ds


def _cloud_run(ds, run_id):
    """One cloud run OF THIS VIDEO DATASET, active or not, or LookupError.
    Ownership is the (id, table) pair — the face lane shares the id space."""
    try:
        run = db.session.get(CloudTrainingRun, int(run_id))
    except (TypeError, ValueError):
        run = None
    if run is None or not crd.owns(run, ds.id, crd.VIDEO):
        raise LookupError('video training run not found')
    return run


def _deployed_index() -> dict:
    """``basename.lower() -> LoraLoader-form name`` of every video LoRA ComfyUI
    can load, from the Studio's own scan — so "deployed" here means exactly
    what the Studio's picker means by it."""
    return {os.path.basename(e['filename']).lower(): e['filename']
            for e in vts.deployed_loras()}


def _undeployable(deployed_as) -> bool:
    """Only a copy sitting in the app's OWN deployment folder is ours to move
    out. A LoRA the user dropped by hand under `h3/` counts as deployed (the
    Studio lists it) but is not something a click in this list may trash."""
    if not deployed_as:
        return False
    sub = os.path.dirname(os.path.normpath(str(deployed_as)))
    return os.path.normcase(sub) == os.path.normcase(os.path.normpath(vts.LORA_SUBDIR))


def _file_row(name, path, deployed) -> dict:
    try:
        size = os.path.getsize(path)
    except OSError:
        size = None
    deployed_as = deployed.get(name.lower())
    return {'filename': name, 'size': size, 'deployed_as': deployed_as,
            'undeployable': _undeployable(deployed_as)}


def _step_rows(steps, paths_of, deployed) -> list:
    """The step groups with each file described (size, deployed state) and the
    step-level `deployed` flag: true only when EVERY file of the step is in
    ComfyUI — half a pair deployed is not a deployed LoRA."""
    out = []
    for s in steps:
        files = [_file_row(n, paths_of(n), deployed) for n in s['files']]
        out.append({'step': s['step'], 'final': bool(s['final']), 'files': files,
                    'deployed': bool(files) and all(f['deployed_as'] for f in files)})
    return out


def _local_saves(ds) -> dict:
    """``{filename: absolute path}`` of this dataset's LOCAL run saves — the
    lane's own listing (`list_run_checkpoints` reads the save root), keyed the
    way `run_checkpoint_files` keys a cloud run's, so both lanes feed one
    grouping."""
    try:
        entries = vtl.list_run_checkpoints(ds.id, ds.user_id)
    except RuntimeError:
        # No local trainer configured on this install (ai-toolkit's output
        # dir is unknown): the local lane simply has no saves to show. The
        # cloud half of the list must not die with it.
        return {}
    return {e['filename']: e['path'] for e in entries}


def local_group(ds, deployed=None) -> dict | None:
    """The local run's saves grouped by step, or None when it has none.

    The FINAL save of a local run is reported with `step: None`: ai-toolkit's
    naming carries no number for it, and unlike a cloud run the local lane
    stamps no `steps` parameter anywhere this list can read without opening
    the run's config — so the label says "Final" and invents no number."""
    saves = _local_saves(ds)
    if not saves:
        return None
    if deployed is None:
        deployed = _deployed_index()
    steps = _group_saves_by_step(saves, target=None)
    return {
        'run_name': vtl.local_run_name(ds),
        'folder': str(vtl.save_root(ds)),
        'active': bool(vtl.video_training_progress(ds.id, ds.user_id)['active']),
        'steps': _step_rows(steps, saves.get, deployed),
    }


def cloud_groups(ds, deployed=None) -> list:
    """DIVERGENCE 4 — upstream lists this dataset's rented-pod runs and the
    steps each brought back. This fork trains video LOCALLY only, so there is
    never a pod harvest to describe. The key stays in `list_checkpoints`'
    answer, empty, because the section renders both lanes from one payload and
    a missing key would be a different shape rather than an empty lane."""
    return []


def list_checkpoints(user_id, dataset_id) -> dict:
    """Everything the Checkpoints & LoRAs section renders, in one answer.

    `can_deploy` is whether ComfyUI has a loras root at all on this install —
    the one fact that turns every 📦 into a stated refusal rather than a click
    that fails."""
    from . import comfy_model_paths
    ds = _dataset(user_id, dataset_id)
    deployed = _deployed_index()
    return {
        'local': local_group(ds, deployed),
        'cloud': cloud_groups(ds, deployed),
        'can_deploy': bool(comfy_model_paths.search_roots('loras')),
        'deploy_folder': vts.LORA_SUBDIR.replace(os.sep, '/'),
        'delete_mode': DELETE_MODE,
    }


def _step_files(ds, run_id, step, final) -> list:
    """``[(filename, path)]`` of ONE step of the local run (`run_id` None) or of
    one cloud run of this dataset. A step that is not there is a LookupError —
    a deploy or a delete must never act on "the nearest" save."""
    # DIVERGENCE 4 — upstream branches here on `run_id`: None is the local run,
    # a number is one of this dataset's rented-pod runs. There is no pod lane
    # here, so a step is always the local run's; a caller naming a run id is
    # asking for something this build cannot have.
    if run_id is not None:
        raise LookupError('unknown checkpoint step')
    saves = _local_saves(ds)
    steps = _group_saves_by_step(saves, target=None)
    paths = saves
    final = bool(final)
    for s in steps:
        if bool(s['final']) != final:
            continue
        if not final and (step is None or int(s['step']) != int(step)):
            continue
        return [(n, paths[n]) for n in s['files'] if n in paths]
    raise LookupError('unknown checkpoint step')


def deploy_step(user_id, dataset_id, run_id, step, final=False) -> dict:
    """📦 Put EVERY file of one step into ComfyUI's loras folder, through the
    Studio's own copy. Answers the LoraLoader-form names it now goes by."""
    ds = _dataset(user_id, dataset_id)
    files = _step_files(ds, run_id, step, final)
    if not files:
        raise LookupError('unknown checkpoint step')
    return {'deployed': [vts.deploy_file(path) for _, path in files],
            'folder': vts.LORA_SUBDIR.replace(os.sep, '/')}


def undeploy(user_id, dataset_id, deployed_as) -> dict:
    """⏏ Move one deployed copy out of ComfyUI (to the trash). The training
    save is untouched — the step offers to deploy again right after."""
    _dataset(user_id, dataset_id)
    return {'removed': vts.undeploy_lora(deployed_as), 'delete_mode': DELETE_MODE}


def delete_step(user_id, dataset_id, run_id, step, final=False) -> dict:
    """🗑 Move every file of one step to the trash — all of a Wan pair, never
    half. Refused while the lane that writes these files is still writing:
    a local training in progress, or a cloud run still on its pod (its sync
    would just bring the file back).

    A file the OS holds open (an antivirus scan, a loader) stays, is counted in
    `files_kept` and named — the answer says what is on disk, not what was
    asked for. The same contract as removing a clip."""
    ds = _dataset(user_id, dataset_id)
    if run_id is None:
        if vtl.video_training_progress(ds.id, ds.user_id)['active']:
            raise RuntimeError('training is running on this set and still writing '
                               'these saves — stop it first')
        run = None
    else:
        run = _cloud_run(ds, run_id)
        if run.status in ct.ACTIVE_STATES:
            raise RuntimeError('this cloud run is still active — its save would '
                               'just be re-synced; stop the run first')
    files = _step_files(ds, run_id, step, final)
    removed, kept = [], []
    for name, path in files:
        try:
            if run is None:
                trash.send_to_trash(path, context=f'videockpt_ds{ds.id}')
            else:
                ct.delete_cloud_checkpoint(ds.id, run.id, name, dataset_table=crd.VIDEO)
        except trash.TrashLockError:
            kept.append(name)
            continue
        except ValueError:
            # Gone between the listing and the click — nothing left to move.
            continue
        removed.append(name)
    return {'removed': removed, 'files_kept': kept, 'delete_mode': DELETE_MODE}


def local_checkpoint_path(user_id, dataset_id, filename) -> str | None:
    """Resolve ONE local save by basename, or None. Resolved through the run's
    own listing, so a request can never point this at a path of its choosing —
    the cloud download's rule (`run_checkpoint_path`), applied to the folder."""
    ds = _dataset(user_id, dataset_id)
    if not filename or os.path.basename(filename) != filename:
        return None
    path = _local_saves(ds).get(filename)
    return path if path and os.path.isfile(path) else None


def run_details(user_id, dataset_id, run_id) -> dict:
    """ⓘ What one cloud run of this dataset was: its status line, its timing,
    its GPU and price, its genealogy, and the parameters it was launched with
    (allow-listed — see `_PUBLIC_PARAMS`)."""
    ds = _dataset(user_id, dataset_id)
    run = _cloud_run(ds, run_id)
    try:
        params = json.loads(run.train_params or '{}')
    except ValueError:
        params = {}
    return {
        'run_id': run.id, 'status': run.status,
        'phase_detail': run.phase_detail or '', 'error': run.error,
        'run_name': run.run_name, 'gpu': run.gpu_name,
        'price_per_hour': run.price_per_hour,
        'created_at': run.created_at.isoformat() if run.created_at else None,
        'finished_at': run.finished_at.isoformat() if run.finished_at else None,
        'parent_run_id': params.get('parent_run_id'),
        'saves': len(ct.run_checkpoint_files(run)),
        'params': {k: params[k] for k in _PUBLIC_PARAMS if k in params},
    }

File: frontend/src/components/videobank/videoCheckpoints.js
/* 📦 Pure brain of the video workspace's Checkpoints & LoRAs section — every
   decision the list makes about a SAVE, JSX-free so `node --test` can pin them.

   The image lane's checkpoint popover (components/dataset/checkpointPopover.js)
   decides per FILE; this one decides per STEP, and that is the whole reason it
   is a second file rather than a second caller of the first: a Wan 2.2
   checkpoint is two files at one step, and a per-file model offers half a LoRA
   without noticing. Same verbs, same rule — an action is live, or shown with
   its real reason, or absent; never a button that fails in silence — same
   wording where the behaviour is the same, and a DIFFERENT sentence where it is
   not (a local video run cannot pick a step to continue from; see
   CONTINUE_LOCAL_REASON), which is CLAUDE.md's parity rule made literal. */
import { deleteDestination, isRecoverable } from '../../utils/deletionWording.js'

/** How one step of saves is named in the list and in every confirm.
 *
 * DIVERGENCE 4 — upstream keeps this in `videoCloudStatus.js`, the module that
 * describes a rented-pod run's phases. That module is not carried here, and
 * this function has nothing to do with a pod: it names a save, and its own rule
 * below is about a LOCAL run. It lives with the local checkpoint consumers. */
export function stepLabel(step) {
  if (!step) return ''
  const n = step.files?.length || 0
  // A LOCAL run's final save carries no number (the lane stamps no step count
  // the listing can read): "Final", and never "Final (step null)".
  const head = step.final
    ? (step.step != null ? `Final (step ${step.step})` : 'Final')
    : `Step ${step.step}`
  return n > 1 ? `${head} — ${n} files (both experts)` : head
}
import { videoDatasetCheckpointUrl, videoDatasetLocalCheckpointUrl } from './videoBankApi.js'

export const EMPTY_NOTE = 'No checkpoints yet — train this set on this PC, and every '
  + 'save appears here, step by step.'

/* Why a LOCAL save offers no ▶ Continue: ai-toolkit resumes from whatever it
   finds in the run folder, so the next local launch continues from the NEWEST
   save whatever row the click came from. Offering the button would do
   something other than what it says. */
export const CONTINUE_LOCAL_REASON = 'Resumes from its newest save on the next '
  + 'local launch — the run folder is the resume state, so no step is picked here'
export const ACTIVE_CLOUD_REASON = 'This run is still on its pod — stop it first'
export const ACTIVE_LOCAL_REASON = 'Training is running and still writing these '
  + 'saves — stop it first'
export const NO_LORAS_ROOT_REASON = 'ComfyUI\'s loras folder is not configured — '
  + 'nothing to deploy into'
export const HAND_PLACED_REASON = 'Deployed by hand — remove it from ComfyUI\'s '
  + 'loras folder yourself'

/** The groups the section renders, in order: the local run first (it is this
 * machine), then the cloud runs newest first — the server's own order. A group
 * with no step is not a group: the empty state is one sentence, not a header
 * over nothing. */
export function checkpointGroups(payload) {
  const out = []
  const local = payload?.local
  if (local?.steps?.length) {
    out.push({
      key: 'local', lane: 'local', run_id: null, active: !!local.active,
      run_name: local.run_name, folder: local.folder, parent_run_id: null,
      status: local.active ? 'training' : 'done', steps: local.steps,
    })
  }
  for (const g of payload?.cloud || []) {
    if (g?.steps?.length) out.push({ key: `cloud-${g.run_id}`, lane: 'cloud', ...g })
  }
  return out
}

export function groupTitle(group) {
  if (group.lane === 'local') return `On this PC — ${group.run_name || 'local run'}`
  const from = group.parent_run_id ? ` — continued from #${group.parent_run_id}` : ''
  return `Cloud run #${group.run_id}${from}`
}

// Relative "15m ago" from a naive-UTC backend timestamp — the Runs page's rule,
// so a group header reads exactly like its Runs row.
export function timeAgo(iso, now = Date.now()) {
  if (!iso) return ''
  const t = new Date(/[Z+]/.test(iso) ? iso : `${iso}Z`).getTime()
  if (Number.isNaN(t)) return ''
  const s = Math.max(0, (now - t) / 1000)
  if (s < 60) return 'just now'
  if (s < 3600) return `${Math.floor(s / 60)}m ago`
  if (s < 86400) return `${Math.floor(s / 3600)}h ago`
  return `${Math.floor(s / 86400)}d ago`
}

export function groupSub(group, now = Date.now()) {
  if (group.lane === 'local') {
    return group.active ? 'training now — saves still being written' : 'this machine\'s run folder'
  }
  const bits = [group.status || 'unknown']
  if (group.gpu) bits.push(group.gpu)
  if (group.price_per_hour != null) bits.push(`$${Number(group.price_per_hour).toFixed(2)}/h`)
  const when = timeAgo(group.finished_at || group.created_at, now)
  if (when) bits.push(when)
  return bits.join(' · ')
}

export function fmtSize(bytes) {
  const n = Number(bytes)
  if (!Number.isFinite(n) || n <= 0) return ''
  if (n >= 1024 ** 3) return `${(n / 1024 ** 3).toFixed(1)} GB`
  if (n >= 1024 ** 2) return `${Math.round(n / 1024 ** 2)} MB`
  return `${Math.max(1, Math.round(n / 1024))} KB`
}

/** The short name a ⬇ link shows: the expert of a Wan pair, else the file. */
export function fileShortName(filename, fileCount = 1) {
  const name = String(filename || '')
  if (fileCount > 1) {
    if (/_high_noise\.safetensors$/i.test(name)) return 'high noise'
    if (/_low_noise\.safetensors$/i.test(name)) return 'low noise'
  }
  return name
}

export function stepKey(group, step) {
  return `${group.key}:${step.final ? 'final' : step.step}`
}

/** One ⬇ per FILE — both halves of a pair side by side is what the loaders
 * expect; the URL comes from the lane the save belongs to. */
export function downloadLinks(datasetId, group, step) {
  const files = step.files || []
  return files.map((f) => ({
    filename: f.filename,
    short: fileShortName(f.filename, files.length),
    size: f.size,
    url: group.lane === 'cloud'
      ? videoDatasetCheckpointUrl(datasetId, group.run_id, f.filename)
      : videoDatasetLocalCheckpointUrl(datasetId, f.filename),
  }))
}

/**
 * Everything a step's row renders, decided in one place.
 *
 *   { key, label, files, deployed,
 *     continue: {ok:true}|{reason},
 *     deploy:   {ok:true, folder}|{reason}|null   (null when deployed),
 *     undeploy: {ok:true}|{reason}|null           (null when not deployed),
 *     del:      {ok:true, label, title}|{reason},
 *     details:  bool }
 *
 * `ctx.deleteMode` is the server's `delete_mode`: the 🗑 title names the real
 * destination through the app-wide wording, never a "trash" typed here.
 */
export function stepActionModel(datasetId, group, step, ctx = {}) {
  const { canDeploy = true, deployFolder = 'h3/lds', deleteMode = 'app_trash' } = ctx
  const files = step.files || []
  const deployed = !!step.deployed && files.length > 0
  const cloud = group.lane === 'cloud'

  let cont
  if (!cloud) cont = { reason: CONTINUE_LOCAL_REASON }
  else if (group.active) cont = { reason: ACTIVE_CLOUD_REASON }
  else cont = { ok: true }

  let deploy = null
  let undeploy = null
  if (deployed) {
    undeploy = files.every((f) => f.undeployable) ? { ok: true } : { reason: HAND_PLACED_REASON }
  } else {
    deploy = canDeploy ? { ok: true, folder: deployFolder } : { reason: NO_LORAS_ROOT_REASON }
  }

  let del
  if (group.active) del = { reason: cloud ? ACTIVE_CLOUD_REASON : ACTIVE_LOCAL_REASON }
  else {
    del = {
      ok: true,
      label: files.length > 1 ? 'Delete the training saves' : 'Delete the training save',
      title: `Move every file of this step to ${deleteDestination(deleteMode)}`
        + (isRecoverable(deleteMode) ? ' — recoverable until you empty it' : ''),
    }
  }

  return {
    key: stepKey(group, step), label: stepLabel(step),
    files: downloadLinks(datasetId, group, step),
    deployed, continue: cont, deploy, undeploy, del, details: cloud,
  }
}

const quoted = (files) => (files || []).map((f) => `“${f.filename}”`).join(' + ')

/** The 🗑 confirmation. Names every file the click moves and the destination —
 * from the app-wide wording, never a sentence of this file's own. */
export function describeStepDelete(group, step, mode) {
  const files = step.files || []
  const many = files.length > 1
  const where = deleteDestination(mode)
  const lines = [
    `DELETE THE TRAINING SAVE${many ? 'S' : ''} — ${quoted(files)} (${stepLabel(step)})?`, '',
    many
      ? 'These are the run\'s own checkpoint files — both experts of the pair go together, never half.'
      : 'This is the run\'s own checkpoint file, not a ComfyUI copy.',
    isRecoverable(mode)
      ? `${many ? 'They go' : 'It goes'} to ${where} — recoverable until you empty it.`
      : `${many ? 'They go' : 'It goes'} to ${where}.`,
  ]
  if (deployed(step)) {
    lines.push('', 'The copy deployed into ComfyUI is a separate file and is KEPT — use ⏏ Undeploy for that one.')
  }
  return lines.join('\n')
}

const deployed = (step) => !!step?.deployed

export function describeUndeploy(step, mode = 'app_trash') {
  const files = (step.files || []).filter((f) => f.deployed_as)
  return [
    `UNDEPLOY — REMOVE FROM COMFYUI — ${quoted(files)} (${stepLabel(step)})?`, '',
    `Only the copy in ComfyUI's loras folder goes to ${deleteDestination(mode)}.`,
    'The training save is KEPT — this step offers to deploy again right after.',
  ].join('\n')
}

/** The run-level 🗑 the training block used to carry, moved here unchanged:
 * this one removes the run's files and its history line for good (the server
 * deletes the store directory by name — no trash), and says so. */
export function runDeleteConfirmation(group) {
  const n = (group.steps || []).reduce((sum, s) => sum + (s.files?.length || 0), 0)
  return `Delete run #${group.run_id} and its ${n} LoRA file(s) from disk?\n\n`
    + 'The dataset and its clips are untouched — only this run’s '
    + 'checkpoints and its history line go. This cannot be undone.'
}

export function deleteReport(res = {}) {
  const removed = res.removed || []
  const kept = res.files_kept || []
  const where = deleteDestination(res.delete_mode)
  const head = removed.length
    ? `Moved ${removed.length} file${removed.length === 1 ? '' : 's'} to ${where}.`
    : 'Nothing was moved.'
  if (!kept.length) return head
  return `${head} ${kept.length} file${kept.length === 1 ? '' : 's'} kept — held open by `
    + `another program: ${kept.join(', ')}.`
}

export function deployReport(res = {}) {
  const names = (res.deployed || []).map((n) => String(n).split(/[\\/]/).pop())
  return `Deployed → ${res.folder || 'h3/lds'}: ${names.join(' + ')}. The Video Test Studio lists it now.`
}

export function undeployReport(step) {
  const files = (step?.files || []).filter((f) => f.deployed_as)
  return `Removed from ComfyUI: ${files.map((f) => f.filename).join(' + ')}. The training save is kept.`
}

/** The ▶ Continue body. `from_step` is the harvested step the server seeds the
 * new pod with — for a FINAL save that is the run's total step count, which is
 * the number the listing reports it at. */
export function continueBody(group, step, extraSteps) {
  const extra = Math.max(1, Math.floor(Number(extraSteps) || 0))
  return { run_id: group.run_id, extra_steps: extra, from_step: step.step }
}

const fmtWhen = (iso) => {
  if (!iso) return null
  const t = new Date(/[Z+]/.test(iso) ? iso : `${iso}Z`)
  return Number.isNaN(t.getTime()) ? String(iso) : t.toLocaleString()
}

/** ⓘ label/value rows of one cloud run, in reading order; a value the run
 * does not carry is not a row. */
export function detailsRows(d = {}) {
  const p = d.params || {}
  const yesNo = (v) => (v == null ? null : (v ? 'yes' : 'no'))
  const rows = [
    ['Status', [d.status, d.phase_detail].filter(Boolean).join(' — ') || null],
    ['GPU', d.gpu ? `${d.gpu}${d.price_per_hour != null ? ` · $${Number(d.price_per_hour).toFixed(2)}/h` : ''}` : null],
    ['Requested GPU class', p.requested_gpu ?? null],
    ['Started', fmtWhen(d.created_at)],
    ['Finished', fmtWhen(d.finished_at)],
    ['Continued from', d.parent_run_id != null ? `run #${d.parent_run_id}` : null],
    ['Resumed at step', p.resume_step ?? null],
    ['Steps', p.steps ?? null],
    ['Target', p.target_profile ?? null],
    ['Frames per clip', p.frames ?? null],
    ['Image-to-video', yesNo(p.do_i2v)],
    ['Low VRAM', yesNo(p.low_vram)],
    ['Distillation', p.distillation ?? null],
    ['Base model', p.base_model || null],
    ['Sample prompts', Array.isArray(p.sample_prompts) ? String(p.sample_prompts.length) : null],
    ['Saves on this machine', d.saves ?? null],
    ['Error', d.error || null],
  ]
  return rows.filter(([, v]) => v != null && v !== '').map(([k, v]) => [k, String(v)])
}
